Mage/src/main/java/mage/abilities/AbilityImpl.java

```diff
@@ -1703,6 +1701,26 @@ public void setSourceObjectZoneChangeCounter(int sourceObjectZoneChangeCounter)
         this.sourceObjectZoneChangeCounter = sourceObjectZoneChangeCounter;
     }
 
+    @Override
+    public void initSourceObjectZoneChangeCounter(Game game, boolean force) {
+        if (!(this instanceof MageSingleton) && (force || sourceObjectZoneChangeCounter == 0 )) {
+            setSourceObjectZoneChangeCounter(getCurrentSourceObjectZoneChangeCounter(game));
+        }
+    }
+
+    private int getCurrentSourceObjectZoneChangeCounter(Game game){
+        int zcc = game.getState().getZoneChangeCounter(getSourceId());
+        if (game.getPermanentEntering(getSourceId()) != null){
+            // If the triggered ability triggered while the permanent is entering the battlefield
+            // then add 1 zcc so that it triggers as if the permanent was already on the battlefield
+            // So "Enters with counters" causes "Whenever counters are placed" to trigger with battlefield zcc
+            // Particularly relevant for Sagas, which always involve both
+            // Note that this does NOT apply to "As ~ ETB" effects, those still use the stack zcc
+            zcc += 1;
```

Comment:
@ssk97 
1. Add new tests from that PR to master
2. Add shared zcc init code to master with disabled zcc+1 and ref comments to PR
3. Disable failed tests and mark it with TODO like susucre did before with ref to PR
4. Rebase that PR to actual version, so it will contains only new zcc + 1, test enabling and exile zone fixes in some cards;